### ISSUE TITLE
(🐞) `detekt` task only checks `src/main/kotlin`, doesn't check the actual source dirs

### ISSUE THREAD

**Issue by KotlinIsland** [detekt/detekt#6669]
This also affects KMP projects, that don't use `src/main`.
```kt
plugins {
    kotlin("jvm") version "2.0.0-Beta1"
    id("io.gitlab.arturbosch.detekt") version "1.23.4"
}

repositories {
    mavenCentral()
}

kotlin {
    sourceSets.main {
        kotlin.setSrcDirs(setOf("src/mainn")) // contains the actual source code
    }
}
```
```console
> gradle detekt --info

> Task :detekt NO-SOURCE
Skipping task ':detekt' as it has no source files and no previous output files.

BUILD SUCCESSFUL in 1s
```

- Related: https://github.com/detekt/detekt/discussions/6662

**Comment by cortinico** [comment]
You should be using the other `detekt*` tasks and not the plain `detekt` task.
See it here: https://detekt.dev/docs/1.23.1/gettingstarted/type-resolution/#enabling-on-an-android-project

**Comment by KotlinIsland** [comment]
@cortinico but what if the situation is like the OP?


Additionally, in a KMP project, detekt doesn't add the KMP tasks the the `check` task.

**Comment by cortinico** [comment]
> It seems like highly desirable behavior to me.

What is the desiderable behavior for you? As it's not clear specifically what you're asking

**Comment by KotlinIsland** [comment]
> What is the desiderable behavior for you? As it's not clear specifically what you're asking

As per the title of the issue, I would like `deteck` to check the actual source dirs, not the hard-coded path `src/main/kotlin`.

**Comment by cortinico** [comment]
> @cortinico Is this a technical limitation? other plugins have no issue using the configured source sets.

It's not a technical limitation. It's part of the historical setup of our plugin. 

We do have a `detekt` task which doesn't use type resoution, and inspects only `src/main/kotlin` and other folders you can manually specify.
Then we do add `detektMain`, `detektTest`, etc. that are generated starting from the srcSet definition, similarly to what ktlint-gradle does

**Comment by KotlinIsland** [comment]
> It's part of the historical setup of our plugin.

Is there any intention to change it? Not sure what 'type resolution' means in this context, but I'm going to assume it means using the configured source sets to find the source code. 

This issue is requesting that the `detekt` task uses the configured source sets, and not a hard coded path. Is there an existing issue to track using type resolution to use the configured source sets? if not then I would like to see this issue reopened.

**Comment by cortinico** [comment]
> Not sure what 'type resolution' means in this context, but I'm going to assume it means using the configured source sets to find the source code.

Explained here: https://detekt.dev/docs/1.23.1/gettingstarted/type-resolution/#what-is-type-resolution

> This issue is requesting that the detekt task uses the configured source sets, and not a hard coded path. Is there an existing issue to track using type resolution to use the configured source sets? if not then I would like to see this issue reopened.

We've been discussing a restructure of the Gradle plugin to make this "simpler". Sadly it's not as easy as just "detekt should use the configured source sets" as type resolution comes with a performance cost. There are valid use cases (i.e. pre-commit hook) where folks wants to run detekt without type resolution. Therefore, the `detekt` task is valid.

The most up to date status of this work is here:
https://github.com/detekt/detekt/discussions/6035#discussioncomment-5702529

Specifically see "Have a `detetkLite/detetkFull` task that will take care of running with/without TR".
In that context, both those new tasks would work as "umbrella" tasks that will invoke the task on your source set (so `detektLite` will invoke `detektMainLite` and `detektTestLite` and `detektFull` will invoke  `detektMainFull` and `detektTestFull`, etc.)

**Comment by KotlinIsland** [comment]
> Then we do add `detektMain`, `detektTest`, etc. that are generated starting from the srcSet definition, similarly to what ktlint-gradle does

The `detektMain` etc don't use the configured source set directory. If I add a new sourceSet: 
```kt
kotlin {
 sourceSets {
     create("other") {
         kotlin.srcDirs("src/other/kotlin")
     }
  }
}
```

Then there are no corresponding detect tasks generated. (`detectOther`/`detectBaselineOther`)

Using the configured source directory has nothing to do with 'type resolution', and the thread you linked doesn't mention supporting it either.

Maybe I am really misunderstanding the situation here, but it really seems like this is a missing feature. Could this issue please be reopened, or an explanaint given as to why this feature won't be supported?
